TaskHero/AppDelegate.swift

```diff
@@ -64,5 +62,65 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         return storyboard.instantiateViewController(withIdentifier: "LoginViewController") as! LoginViewController
     }
     
+    private func registerForLogoutMessages() {
+        NotificationCenter.default.addObserver(forName: NSNotification.Name(rawValue: User.didLogoutNotification), object: nil, queue: OperationQueue.main, using: {(notification: Notification) -> Void in
+            self.setRootViewController(self.loadLoginScreen())
+        })
+    }
+    
+}
+
+// MARK: functions related to push
+
+extension AppDelegate: UNUserNotificationCenterDelegate {
+    
+    func clearBadges() {
+        let installation = PFInstallation.current()
+        installation?.badge = 0
+        installation?.saveInBackground { (success, error) -> Void in
+            if success {
+                UIApplication.shared.applicationIconBadgeNumber = 0
+            }
+            else {
+                NSLog("Failed to clear badges")
+            }
+        }
+    }
+    
+    func setupNotifications(_ application: UIApplication) {
+        let center  = UNUserNotificationCenter.current()
+        center.delegate = self
+        center.requestAuthorization(options: [.sound,.alert,.badge]) { (granted, error) in
+            if granted {
+                application.registerForRemoteNotifications()
+            } else {
+                NSLog("Error requesting push notification authorization")
+            }
+        }
+    }
+    
+    func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
+        let token = deviceToken.map { String(format: "%02.2hhx", $0) }.joined()
+        NSLog("Registration succeeded! Token: \(token)")
+        
+        let installation = PFInstallation.current()
+        installation?.setDeviceTokenFrom(deviceToken)
+        installation?.channels = ["global"]
+        installation?.saveInBackground()
+    }
+    
+    func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
+        NSLog("Push notifications registration failed; note that this feature not supported in the iOS Simulator, so use your phone if so.")
+    }
+    
+    func userNotificationCenter(_ center: UNUserNotificationCenter,  willPresent notification: UNNotification, withCompletionHandler   completionHandler: @escaping (_ options:   UNNotificationPresentationOptions) -> Void) {
+            PFPush.handle(notification.request.content.userInfo)
+            print("\(notification.request.content.userInfo)")
+    }
+    
+    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
```

Comment:
I don't think this delegate method is useful to us since it involves custom actions ([see here](https://developer.apple.com/reference/usernotifications/unusernotificationcenterdelegate/1649501-usernotificationcenter)), which I don't think we're utilizing in the push notifs. That would be like if the user dragged down the notification and we had some pre-made responses that said "Mark completed" "reply" etc. I don't think there's harm in leaving it in...mostly just don't want to be confused between this and the other delegate that receives notifs when the app is in the foreground